consensus/types.go

```diff
@@ -197,29 +222,78 @@ func (b Block) MinerPayoutID(i int) OutputID {
 	return OutputID(hash.HashBytes(encoding.MarshalAll(b.ID(), i)))
 }
 
-// FileContractID returns the id of a file contract given the index of the contract.
-//
-// TODO: Reconsider how file contract ids are derived
+// FileContractID returns the id of a file contract given the index of the
+// contract. The id is derived by marshalling all of the fields in the
+// transaction except for the signatures and then appending the string "file
+// contract" and the index of the contract.
 func (t Transaction) FileContractID(i int) ContractID {
 	return ContractID(hash.HashBytes(encoding.MarshalAll(
-		t.Outputs[0],
-		t.FileContracts[i],
-		"contract",
+		FileContractIdentifier,
+		t.Inputs,
+		t.MinerFees,
+		t.Outputs,
+		t.FileContracts,
+		t.StorageProofs,
+		t.SiafundInputs,
+		t.SiafundOutputs,
+		t.ArbitraryData,
```

Comment:
now we can safely use `hash.HashAll`... except it doesn't exist.

This should work though; it avoids the double-interface problem:
```go
func HashAll(objs ...interface{}) Hash {
    var b[]byte
    for _, obj := range objs {
        b = append(b, encoding.Marshal(obj))
    }
    return HashBytes(b)
}
```